src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -49924,6 +49924,240 @@ static object F((I<object>, I<object?>)? x, (I<object?>, I<object>) y)
+            var source =
+@"using System;
+Action<string?> a1 = param => {};
+Action<string?> a2 = a1 ?? (Action<string>)((string s) => {});
+";
+
+            var comp = CreateCompilation(source, options: WithNullableEnable(TestOptions.ReleaseExe));
+            comp.VerifyDiagnostics(
+                // (3,22): warning CS8619: Nullability of reference types in value of type 'Action<string>' doesn't match target type 'Action<string?>'.
+                // Action<string?> a2 = a1 ?? (Action<string>)((string s) => {});
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "a1 ?? (Action<string>)((string s) => {})").WithArguments("System.Action<string>", "System.Action<string?>").WithLocation(3, 22)
+            );
+        }
+
+        [Fact, WorkItem(29955, "https://github.com/dotnet/roslyn/issues/29955")]
+        public void NullCoalescingOperator_16()
+        {
+            var source =
+@"using System;
+Func<string> a1 = () => string.Empty;
+Func<string> a2 = a1 ?? (() => null);
+";
+
+            var comp = CreateCompilation(source, options: WithNullableEnable(TestOptions.ReleaseExe));
+            comp.VerifyDiagnostics(
+                // (3,32): warning CS8603: Possible null reference return.
+                // Func<string> a2 = a1 ?? (() => null);
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(3, 32)
+            );
+        }
+
+        [Fact, WorkItem(29955, "https://github.com/dotnet/roslyn/issues/29955")]
+        public void NullCoalescingOperator_17()
+        {
+            var source =
+@"using System;
+Func<string> a1 = () => string.Empty;
+Func<string> a2 = a1 ?? (Func<string?>)(() => null);
+";
+
+            var comp = CreateCompilation(source, options: WithNullableEnable(TestOptions.ReleaseExe));
+            comp.VerifyDiagnostics(
+                // (3,19): warning CS8619: Nullability of reference types in value of type 'Func<string?>' doesn't match target type 'Func<string>'.
+                // Func<string> a2 = a1 ?? (Func<string?>)(() => null);
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "a1 ?? (Func<string?>)(() => null)").WithArguments("System.Func<string?>", "System.Func<string>").WithLocation(3, 19)
+            );
+        }
+
+        [Fact, WorkItem(51975, "https://github.com/dotnet/roslyn/issues/51975")]
+        public void NullCoalescingOperator_18()
+        {
+            var source =
+@"using System.Diagnostics.CodeAnalysis;
+C? c = new C();
+D d1 = c ?? new D();
+d1.ToString();
+
+D d2 = ((C?)null) ?? new D();
+d2.ToString();
+
+c = null;
+D d3 = c ?? new D();
+d3.ToString();
+
+class C {}
+
+class D
+{
+    [return: NotNullIfNotNull(""c"")]
+    public static implicit operator D?(C c) => default!;
+}
+";
+
+            var comp = CreateCompilation(new[] { source, NotNullIfNotNullAttributeDefinition }, options: WithNullableEnable(TestOptions.ReleaseExe));
+            comp.VerifyDiagnostics();
+        }
```

Comment:
Is there a test where the LHS has maybe-null flow state, but its conversion is a user-defined conversion that requires non-null input? (expect no diagnostic for this since the LHS is only converted if not null)

Reply:
`NullCoalescingOperator_18`

Reply:
😅